app/src/main/java/com/hedvig/app/feature/marketpicker/MarketPickerActivity.kt

```diff
@@ -0,0 +1,117 @@
+import com.hedvig.app.makeLocaleString
+import com.hedvig.app.util.extensions.compatDrawable
+import com.hedvig.app.util.extensions.getStoredBoolean
+import com.hedvig.app.util.extensions.observe
+import com.hedvig.app.util.extensions.storeBoolean
+import com.hedvig.app.util.extensions.view.setHapticClickListener
+import kotlinx.android.synthetic.main.activity_market_picker.*
+import org.koin.android.ext.android.inject
+import org.koin.android.viewmodel.ext.android.viewModel
+
+class MarketPickerActivity : BaseActivity(R.layout.activity_market_picker) {
+    private val marketModel: MarketPickerViewModel by viewModel()
+    private val languageViewModel: LanguageViewModel by viewModel()
+    private val tracker: LanguageSelectionTracker by inject()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        storeBoolean(HAS_SHOWN_MARKET_SELECTION, true)
+
+        var marketAdapter: MarketAdapter? = null
+        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
+
+        marketModel.preselectedMarket.observe(this) { marketString ->
+            marketString?.let {
+                val market = Market.valueOf(marketString)
+                marketAdapter = MarketAdapter(marketModel, market.ordinal)
+                countryList.adapter = marketAdapter
+                countryList.addItemDecoration(
+                    DividerItemDecoration(this, DividerItemDecoration.VERTICAL).apply {
+                        compatDrawable(R.drawable.divider)?.let { setDrawable(it) }
+                    }
+                )
+            }
+
+        }
+        marketModel.loadGeo()
+
+        languageList.adapter = LanguageAdapterNew(languageViewModel, Market.SE)
+        marketModel.selectedMarket.observe(this) { market ->
+            market?.let {
+                languageList.adapter = LanguageAdapterNew(languageViewModel, market)
+            }
+        }
+        languageList.addItemDecoration(
+            DividerItemDecoration(this, DividerItemDecoration.VERTICAL).apply {
+                compatDrawable(R.drawable.divider)?.let { setDrawable(it) }
+            }
+        )
+
+        save.setHapticClickListener {
+            marketAdapter?.let {
+                sharedPreferences.edit()
+                    .putInt(Market.MARKET_SHARED_PREF, marketAdapter!!.getSelectedMarket())
+                    .commit()
+            }
+            val language = languageViewModel.selectedLanguage.value
+            language?.let {
+                setLanguage(language, languageViewModel)
+            }
+            goToMarketingActivity()
+        }
+
+        languageViewModel.selectedLanguage.observe(this) { language ->
+            save.isEnabled = language != null
+        }
+    }
+
+    @SuppressLint("ApplySharedPref") // We want to apply this right away. It's important
+    private fun setLanguage(language: Language, languageViewModel: LanguageViewModel) {
```

Comment:
```suggestion
    private fun setLanguage(language: Language) {
```

and yes, I know this is my code, I'm just covering up my past mistakes 😬